### PR TITLE
[PHPDoc] Fix some union type cases

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

While working on https://github.com/symfony/symfony/issues/40154, I discovered some PHPDoc issues, I'm going to comment in the review. Upper branches will need some fixes too.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php [hunk 1]
```diff
@@ -389,7 +389,7 @@ public function getAccessDecisionLog()
     /**
      * Returns the configuration of the current firewall context.
      *
-     * @return array|Data
+     * @return array|Data|null
```
Comment (fancyweb) [inline src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php:392]: ` $this->data['firewall'] = null;` line 194

File: src/Symfony/Component/BrowserKit/Response.php [hunk 2]
```diff
@@ -123,7 +123,7 @@ public function getHeaders()
      * @param string $header The header name
      * @param bool   $first  Whether to return the first value or all header values
      *
-     * @return string|array The first header value if $first is true, an array of values otherwise
+     * @return string|array|null The first header value if $first is true, an array of values otherwise
```
Comment (fancyweb) [inline src/Symfony/Component/BrowserKit/Response.php:126]: `return $first ? null : [];` line 141

File: src/Symfony/Component/Console/Descriptor/Descriptor.php [hunk 3]
```diff
@@ -72,36 +72,26 @@ protected function write($content, $decorated = false)
 
     /**
      * Describes an InputArgument instance.
-     *
-     * @return string|mixed
      */
     abstract protected function describeInputArgument(InputArgument $argument, array $options = []);
```
Comment (fancyweb) [inline src/Symfony/Component/Console/Descriptor/Descriptor.php:76]: We never return anything in the implementations and the class is @internal.

File: src/Symfony/Component/Console/Descriptor/TextDescriptor.php [hunk 4]
```diff
@@ -298,7 +298,7 @@ private function formatDefaultValue($default): string
     }
 
     /**
-     * @param (Command|string)[] $commands
+     * @param array<Command|string> $commands
```
Comment (fancyweb) [inline src/Symfony/Component/Console/Descriptor/TextDescriptor.php:301]: I don't know about this notation, I think `array<Foo|string>` is widespread in SCA tools. I plan to support this notation in the `DebugClassLoader`. Note that here it does not matter on a `@param` but some `@return` use it and I changed them all for consitency.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php [hunk 5]
```diff
@@ -97,7 +97,7 @@ public static function disableStty()
     /**
      * Asks the question to the user.
      *
-     * @return bool|mixed|string|null
+     * @return mixed
```
Comment (fancyweb) [inline src/Symfony/Component/Console/Helper/QuestionHelper.php:100]: `mixed` includes everything in PHP 8. Using `mixed|somethingElse` as a return type is a fatal error.
Obviously, we are going to handle the case in the `DebugClassLoader` when a PHPDoc is "wrong".
But in our own codebase, we should be strict and clean about it, WDYT?

File: src/Symfony/Component/Routing/Loader/XmlFileLoader.php [hunk 6]
```diff
@@ -370,7 +370,7 @@ private function parseDefaultsConfig(\DOMElement $element, string $path)
     /**
      * Recursively parses the value of a "default" element.
      *
-     * @return array|bool|float|int|string The parsed value
+     * @return array|bool|float|int|string|null The parsed value
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Loader/XmlFileLoader.php:373]: `return null;` line 346

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php [hunk 7]
```diff
@@ -97,15 +97,15 @@ public function cancel(): void;
      *  - response_headers (array) - an array modelled after the special $http_response_header variable
      *  - start_time (float) - the time when the request was sent or 0.0 when it's pending
      *  - url (string) - the last effective URL of the request
-     *  - user_data (mixed|null) - the value of the "user_data" request option, null if not set
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/ResponseInterface.php:100]: I think we should keep all `mixed|null` when null is documented with a special behavior

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 5 above)
Comment (Nyholm) [reply src/Symfony/Component/Console/Helper/QuestionHelper.php:100]: I agree that we should be strict and clean. But could be be more specific? 

`mixed` is the same as `string|int|float|bool|null|array|object|callable|resource`. 

I know that `Question::$default` has `mixed`, but that is also wrong. 

Is it reasonable to do anything else than `bool|string|null|int|float`?

File: src/Symfony/Component/Routing/Loader/XmlFileLoader.php (hunk 6 above)
Comment (Nyholm) [reply src/Symfony/Component/Routing/Loader/XmlFileLoader.php:373]: You mean line 380?

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 7 above)
Comment (Nyholm) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php:100]: Hm.. but `null` is included in `mixed`. I dont think `mixed|null` make sense. However, I would be very happy to see a more specific definition. Im sure this `mixed` will not include any `callable`.

**Review by Nyholm — APPROVED** [review]
Thank you @fancyweb

Well done finding and fixing these. I've double checked them all and I think they are all good. Im happy with this PR after we find a good solution to the `mixed|null` thing pointed out by Nicolas.

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php:100]: Actually yes, callable is a very valid type here - this is really `mixed`.
mixed contains null, but here we're not describing types for the engine, but types for humans, and documenting that null means something special is valuable I believe.

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Contracts/HttpClient/ResponseInterface.php (hunk 7 above)
Comment (fabpot) [reply src/Symfony/Contracts/HttpClient/ResponseInterface.php:100]: Using just `mixed` is fine for me. The `null` case is documented in the description anyway.

File: src/Symfony/Component/Console/Helper/QuestionHelper.php (hunk 5 above)
Comment (fancyweb) [reply src/Symfony/Component/Console/Helper/QuestionHelper.php:100]: I did not look at the code here. `mixed` is probably wrong indeed, that's another fix to do.

File: src/Symfony/Component/Routing/Loader/XmlFileLoader.php (hunk 6 above)
Comment (fancyweb) [reply src/Symfony/Component/Routing/Loader/XmlFileLoader.php:373]: Yes sorry :smiley:

**Review by nicolas-grekas — APPROVED** [review]

**Comment by Nyholm** [comment]
Thank you

**Pull request merged**